src/Bicep.LangServer/Snippets/ResourceSnippetsProvider.cs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection;
+
+namespace Bicep.LanguageServer.Snippets
+{
+    public class ResourceSnippetsProvider : IResourceSnippetsProvider
+    {
+        private static readonly Dictionary<string, string> _snippetNameToDetailMap = new Dictionary<string, string>()
```

Comment:
>_snippetNameToDetailMap [](start = 59, length = 23)

Will any tests break if a new snippet is added without a corresponding detail mapping?

Reply:
Yes, it will. Both the tests I modified as part of this PR. I am ok with the integration test. I don't see a lot of value in changes I had to make in unit test - DeclarationSnippetsShouldBeValid. Do you have any suggestions?

Reply:
The original intent of some of those tests was to ensure the snippet contents are valid to some extent. I am wondering if it'd be possible to transform the snippet file content in a generic fashion to produce a syntactically valid bicep file. We can't guarantee zero semantic diagnostics, but could we at least assert on syntax? We do have some pieces (like syntax rewriters) to help with some aspects of this. Probably worth a discussion...

---
In reply to: [595629500](https://github.com/Azure/bicep/pull/1904#discussion_r595629500) [](ancestors = 595629500)

Reply:
I don't think we would need to block the PR on the outcome of the above discussion/brainstorming, though.

---
In reply to: [595633350](https://github.com/Azure/bicep/pull/1904#discussion_r595633350) [](ancestors = 595633350,595629500)

Reply:
Something I'm wondering about - what if we enforced that for each snippet we check in, we had a corresponding test `.bicep` file, which we expect to be 100% compilable without any warnings or errors. It would be really nice to be able to know for certain that we haven't inadvertently made a compiler change causing snippets to be invalid.

Thinking something like the following - where our test understands the `// insert snippet` instruction with replacement values. It could then start with the placeholder file, requests the snippet, inserts and then compiles the resulting file:
```bicep
var aksName = 'test'
... any other prerequisites (dependent resources, child resources etc) go here...

// insert snippet 'res-aks-cluster.bicep'
// $1 = aksName
// $2 = 1.15.7
// ... etc
```

One benefit to simplifying this is that it would lower the bar for taking external contributions.

Reply:
This is a great suggestion. It will make it possible for all possible snippets to compile without any warnings or errors given a corresponding test file.